src/schema/FroidSchema.ts

```diff
@@ -0,0 +1,828 @@
+export type NodeQualifier = (
+  node: ASTNode,
+  objectTypes: ObjectTypeNode[]
+) => boolean;
+
+export type FroidSchemaOptions = {
+  contractTags?: string[];
+  keySorter?: KeySorter;
+  nodeQualifier?: NodeQualifier;
+  typeExceptions?: string[];
+};
+
+/**
+ * The default key sorter
+ * This is used to sort keys in preparation for selecting the FROID key.
+ * This sorter defaults to _no sorting whatsoever_.
+ *
+ * @param {Key[]} keys - The keys
+ * @returns {Key[]} the keys in the order provided
+ */
+const defaultKeySorter: KeySorter = (keys: Key[]): Key[] => keys;
+
+const defaultNodeQualifier: NodeQualifier = () => true;
+
+const scalarNames = specifiedScalarTypes.map((scalar) => scalar.name);
+
+// Custom types that are supported when generating node relay service schema
+const typeDefinitionKinds = [
+  Kind.SCALAR_TYPE_DEFINITION,
+  Kind.ENUM_TYPE_DEFINITION,
+];
+
+/**
+ * A factory for creating FROID schema AST
+ */
+export class FroidSchema {
+  /**
+   * The contract tags that will be applied to FROID schema.
+   */
+  private readonly contractTags: ConstDirectiveNode[];
+  /**
+   * The Apollo Federation version that will be applied to the FROID schema.
+   */
+  private readonly federationVersion: string;
+  /**
+   * The key sorting function.
+   */
+  private readonly keySorter: KeySorter;
+  /**
+   * The node qualifier function.
+   */
+  private readonly nodeQualifier: NodeQualifier;
+  /**
+   * the list of types that should be omitted from the FROID schema.
+   */
+  private readonly typeExceptions: string[];
+  /**
+   * Definition nodes from across the source schemas after eliminating nodes that should be ignored.
+   */
+  private readonly filteredDefinitionNodes: DefinitionNode[];
+  /**
+   * Object type extension and definition nodes from across all the source schemas.
+   */
+  private readonly extensionAndDefinitionNodes: ObjectTypeNode[];
+  /**
+   * Object type definitions from across the source schemas.
+   */
+  private readonly objectTypes: ObjectTypeDefinitionNode[];
+  /**
+   * The object types that should be included in the FROID schema.
+   */
+  private froidObjectTypes: Record<string, ObjectType> = {};
+  /**
+   * The final FROID AST.
+   */
+  private froidAst: DocumentNode;
+
+  /**
+   * Creates the FROID schema.
+   *
+   * @param {string} name - The name of the subgraph that serves the FROID schema.
+   * @param {string} federationVersion - The Apollo Federation version to use for the FROID schema.
+   * @param {Map<string, string>} schemas - The source schemas from which the FROID schema will be generated.
+   * @param {FroidSchemaOptions} options - The options for FROID schema generation.
+   */
+  constructor(
+    name: string,
+    federationVersion: string,
+    schemas: Map<string, string>,
+    options: FroidSchemaOptions
+  ) {
+    this.federationVersion = federationVersion;
+
+    assert(
+      this.federationVersion.indexOf(FED2_VERSION_PREFIX) > -1,
```

Comment:
Would `this.federationVersion.startsWith(FED2_VERSION_PREFIX)` be better here? Since indexOf would return true for something random like `"v3.v2.1"`

Reply:
Nice catch. I'll give that a try.